### DISCUSSION TITLE
up-keep on element inside up-poll

### DISCUSSION THREAD

**Discussion by lcmen** [unpoly/unpoly#365] [Q&A]
Does Unpoly support `up:fragment:keep` event (to block update with `event.preventDefault`) for an element within `up-poll` driven container?

To give you an example, imagine a list of messages with inline edit form like this:

```html
<div id="messages" up-poll up-source="/messages">
  <div id="message-1">
    ...
    <%= link_to "Edit", edit_message_path(message), "up-target" => "#message-1" />
  </div>
  ...
</div>
```

What I would like to achieve is the ability to keep edit form for a single message while all other messages are being updated by the `up-poll`. Is it achievable with Unpoly? So far I came with custom attribute (`[up-hold]`) that can be added to edit view:

```html
<div id="message" up-hold>
  <form ... >...</form>
</div>
```

Where `up-hold` is defined as follows:

```javascript
up.compiler('[up-hold]', function(element) {
  const poller = element.closest('[up-poll]')
  poller && up.radio.stopPolling(poller)

  return () => poller && up.radio.startPolling(poller)
})
```

But it blocks the whole container from being polled. Any ideas? Thanks!

**Comment by triskweline** [comment]
> Does Unpoly support up:fragment:keep event (to block update with event.preventDefault) for an element within up-poll driven container?

Yes. Just set an `[up-keep]` attribute on that element. No need to pause polling.

**Reply by lcmen** [reply]
@triskweline as fair as I understand `[up-keep]` must also be on the element returned from the `up-poll` response.

What about a scenario that I mentioned above where edit form is displayed inline with other messages?

```html
<div id="messages" up-poll up-source="/messages">
  <div id="message-1">
    ...
    <%= link_to "Edit", edit_message_path(message), "up-target" => "#message-1" />
  </div>
  ...
</div>
```

Clicking on the edit button will return HTML like this:

```html
<div id="message-1" up-keep>
  <form ... >...</form>
</div>
```

And it will display the form instead of regular message on the list.

But since `/messages` returns structure from the first snippet (messages without `[up-keep]`), inline form is replaced during poll.

**Reply by triskweline** [reply]
You could give all your message containers an `[up-keep]` attribute, even when they're not in editing mode.

The `[up-poll]` would now reveal new messages from other users, but no longer show updates to an existing message. If this is important you can either implement change detection in an `up:fragment:keep` listener, or make the messages edit time part of the `[up-keep]` selector:

```html
<div id="message-1" updated="2022-03-15 21:19:30" up-keep="#message-1[updated='2022-03-15 21:19:30']">
  ...
</div>
```

This would not keep the fragment if the `[updated]` timestamp changed.

**Reply by lcmen** [reply]
Thank you @triskweline !

I was going through https://www.hotrails.dev/ (Hotwire tutorial), then decided to replace Hotwire with Unpoly to see if I can achieve the same effect (https://github.com/lcmen/hotquote/pull/1/files). I had to add support for flash messages with partial updates and replace websockets updates with polling but I must say I'm impressed with the effect. IMHO code is easier to understand and follow. It's great library!
